### PR TITLE
Add deployment for Stackdriver Metadata Agent with version and resource requirement controlled by env variable

### PR DESCRIPTION
**Pull request by kawych**
**What this PR does / why we need it**:
Introduces Stackdriver Metadata Agent - a daemon set providing metadata for kubernetes objects connected to the same node.

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by piosz** [comment]
@igorpeshansky PTAL

**Comment by piosz** [comment]
@mikedanese PTAL

**Comment by igorpeshansky** [comment]
LGTM, but would be good if @bmoyles0117 also eyeballed this.

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    metadata:
+      labels:
+        app: stackdriver-agents
+    spec:
+      containers:
+      - image: us.gcr.io/container-monitoring-storage/stackdriver-metadata-agent:{{ metadata_agent_version }}
+        imagePullPolicy: IfNotPresent
+        name: metadata-agent
+        ports:
+        - containerPort: 8000
```
Comment (bmoyles0117) [inline cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: nit: as this rolls into production, we will want to look at occupying a less common port.

File: cluster/common.sh [hunk 2]
```diff
@@ -610,6 +610,8 @@ KUBERNETES_MASTER_NAME: $(yaml-quote ${KUBERNETES_MASTER_NAME})
 ALLOCATE_NODE_CIDRS: $(yaml-quote ${ALLOCATE_NODE_CIDRS:-false})
 ENABLE_CLUSTER_MONITORING: $(yaml-quote ${ENABLE_CLUSTER_MONITORING:-none})
 ENABLE_METRICS_SERVER: $(yaml-quote ${ENABLE_METRICS_SERVER:-false})
+ENABLE_METADATA_AGENT: $(yaml-quote ${ENABLE_METADATA_AGENT:-none})
```
Comment (bmoyles0117) [inline cluster/common.sh]: This sounds more like a boolean to me, should it be?

File: cluster/gce/config-default.sh [hunk 3]
```diff
@@ -145,6 +145,16 @@ ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-influxdb}"
...
 
+# Optional: Metadata agent to setup as part of the cluster bring up:
+#   none        - No metadata agent
+#   stackdriver - Stackdriver metadata agent
+# Metadata agent is a daemon set that provides metadata of kubernetes objects
+# running on the same node for exporting metrics and logs.
+ENABLE_METADATA_AGENT="${KUBE_ENABLE_METADATA_AGENT:-none}"
+
+# Version tag of metadata agent
+METADATA_AGENT_VERSION="${KUBE_METADATA_AGENT_VERSION:-0.2-0.0.13-5}"
```
Comment (bmoyles0117) [inline cluster/gce/config-default.sh]: Caution, this image does not incorporate the performance updates using the watch implementation.

File: cluster/gce/container-linux/configure-helper.sh [hunk 4]
```diff
@@ -1289,6 +1289,20 @@ EOF
     sed -i -e "s@{{ *metrics_cpu_per_node *}}@${metrics_cpu_per_node}@g" "${controller_yaml}"
     update-prometheus-to-sd-parameters ${controller_yaml}
   fi
+  if [[ "${ENABLE_CLUSTER_MONITORING:-}" == "stackdriver" ]] ||
+     ([[ "${ENABLE_CLUSTER_LOGGING:-}" == "true" ]] &&
```
Comment (bmoyles0117) [inline cluster/gce/container-linux/configure-helper.sh]: Just to be sure, we've incorporated building the monitored resource directly into heapster for the new resource types, right?

File: cluster/gce/container-linux/configure-helper.sh [hunk 5]
```diff
@@ -1289,6 +1289,20 @@ EOF
...
+     [[ "${LOGGING_DESTINATION:-}" == "gcp" ]]); then
+    if [[ "${ENABLE_METADATA_AGENT:-}" == "stackdriver" ]] &&
+       [[ "${METADATA_AGENT_VERSION:-}" != "" ]]; then
+      metadata_agent_cpu_request="${METADATA_AGENT_CPU_REQUEST:-40m}"
+      metadata_agent_memory_request="${METADATA_AGENT_MEMORY_REQUEST:-50Mi}"
+      setup-addon-manifests "addons" "metadata-agent/stackdriver"
+      deployment_yaml="${dst_dir}/metadata-agent/stackdriver/metadata-agent.yaml"
```
Comment (bmoyles0117) [inline cluster/gce/container-linux/configure-helper.sh]: Small nit: this is a daemon set, not a deployment. Doesn't really matter, but could seem unclear.

File: cluster/gce/config-default.sh (hunk 3 above)
Comment (bmoyles0117) [reply cluster/gce/config-default.sh]: The correct image tag to use here is 0.2-0.0.13-5-watch if we're going to go with the latest version of the metadata agent that implements watch.

File: cluster/common.sh (hunk 2 above)
Comment (kawych) [reply cluster/common.sh]: I prefer to leave it in the same convention as ENABLE_CLUSTER_MONITORING, to support a case where similar agents would be needed for another monitoring system. This is also consistent with addon directory structure introduced in this PR: .../metadata-agent/stackdriver/ as opposed to just .../stackdriver-metadata-agent/

File: cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml (hunk 1 above)
Comment (kawych) [reply cluster/addons/metadata-agent/stackdriver/metadata-agent.yaml]: Thanks for pointing this out, I'll change it in a follow-up PR. Do we need to agree on some port number (e.g. for fluentd to communicate), or it doesn't make a difference for you now?

File: cluster/gce/config-default.sh (hunk 3 above)
Comment (kawych) [reply cluster/gce/config-default.sh]: The purpose of env variable is to overwrite this value without the need of change in OSS code. I can update the tag in a follow up PR, to keep it more up-to-date.

File: cluster/gce/container-linux/configure-helper.sh (hunk 4 above)
Comment (kawych) [reply cluster/gce/container-linux/configure-helper.sh]: Right.

File: cluster/gce/container-linux/configure-helper.sh (hunk 5 above)
Comment (kawych) [reply cluster/gce/container-linux/configure-helper.sh]: Will address in a follow-up PR

**Pull request merged**